src/entity/object/AreaEffectCloud.php

```diff
@@ -0,0 +1,385 @@
+
+		$this->effectContainer = new EffectContainer();
+		$this->effectContainer->getEffectAddHooks()->add(function() : void{ $this->recalculateEffectColor(); });
+		$this->effectContainer->getEffectRemoveHooks()->add(function() : void{ $this->recalculateEffectColor(); });
+
+		$worldTime = $this->getWorld()->getTime();
+		$this->age = $worldTime - $nbt->getLong(self::TAG_SPAWN_TICK, $worldTime);
+		$this->duration = $nbt->getInt(self::TAG_DURATION, self::DURATION);
+		$this->durationOnUse = $nbt->getInt(self::TAG_DURATION_ON_USE, self::DURATION_ON_USE);
+		$this->reapplicationDelay = $nbt->getInt(self::TAG_REAPPLICATION_DELAY, self::REAPPLICATION_DELAY);
+		$this->radius = $nbt->getFloat(self::TAG_RADIUS, self::RADIUS);
+		$this->radiusOnUse = $nbt->getFloat(self::TAG_RADIUS_ON_USE, self::RADIUS_ON_USE);
+		$this->radiusPerTick = $nbt->getFloat(self::TAG_RADIUS_PER_TICK, self::RADIUS_PER_TICK);
+
+		/** @var CompoundTag[]|ListTag|null $effectsTag */
+		$effectsTag = $nbt->getListTag(self::TAG_EFFECTS);
+		if($effectsTag !== null){
+			foreach($effectsTag as $e){
+				$effect = EffectIdMap::getInstance()->fromId($e->getByte("Id"));
+				if($effect === null){
+					continue;
+				}
+
+				$this->effectContainer->add(new EffectInstance(
+					$effect,
+					$e->getInt("Duration"),
+					Binary::unsignByte($e->getByte("Amplifier")),
+					$e->getByte("ShowParticles", 1) !== 0,
+					$e->getByte("Ambient", 0) !== 0
+				));
+			}
+		}
+		$this->recalculateEffectColor();
+	}
+
+	public function saveNBT() : CompoundTag{
+		$nbt = parent::saveNBT();
+
+		$nbt->setLong(self::TAG_SPAWN_TICK, $this->getWorld()->getTime() - $this->age);
+		$nbt->setShort(self::TAG_POTION_ID, PotionTypeIdMap::getInstance()->toId($this->potionType));
+		$nbt->setInt(self::TAG_DURATION, $this->duration);
+		$nbt->setInt(self::TAG_DURATION_ON_USE, $this->durationOnUse);
+		$nbt->setInt(self::TAG_REAPPLICATION_DELAY, $this->reapplicationDelay);
+		$nbt->setFloat(self::TAG_RADIUS, $this->radius);
+		$nbt->setFloat(self::TAG_RADIUS_ON_USE, $this->radiusOnUse);
+		$nbt->setFloat(self::TAG_RADIUS_PER_TICK, $this->radiusPerTick);
+
+		if(count($this->effectContainer->all()) > 0){
+			$effects = [];
+			foreach($this->effectContainer->all() as $effect){
+				$effects[] = CompoundTag::create()
+					->setByte("Id", EffectIdMap::getInstance()->toId($effect->getType()))
+					->setByte("Amplifier", Binary::signByte($effect->getAmplifier()))
+					->setInt("Duration", $effect->getDuration())
+					->setByte("Ambient", $effect->isAmbient() ? 1 : 0)
+					->setByte("ShowParticles", $effect->isVisible() ? 1 : 0);
+			}
+			$nbt->setTag(self::TAG_EFFECTS, new ListTag($effects));
+		}
+
+		return $nbt;
+	}
+
+	public function isFireProof() : bool{
+		return true;
+	}
+
+	public function canBeCollidedWith() : bool{
+		return false;
+	}
+
+	/**
+	 * Recalculates the area entity's potion bubbles colour based on the active effects.
+	 */
+	protected function recalculateEffectColor() : void{
```

Comment:
Why is this logic duplicated? The same already exists in `EffectManager`.

Reply:
Not really, `EffectManager`'s ignores the color if the effect does not have bubbles and the default color is different.

Reply:
That's one detail. It doesn't require you to duplicate the whole section of code.

Reply:
I added `setDefaultBubbleColor` and `setEffectValidatorForBubbles` api methods, I would like to know if you find them good enough to be able to mark this conversation as resolved.